From da7681cf53cbb854f4136e24029bc3bd6169ef5a Mon Sep 17 00:00:00 2001
From: Ievgen Sherstobitov <ievgen.sherstobitov@lge.com>
Date: Wed, 23 Dec 2015 13:21:31 +0200
Subject: [PATCH] Prevent creating /usr/etc in native sysroot

:Release Notes:
Prevent creating /usr/etc directory in native sysroot while
executing "npm install"

:Detailed Notes:
Currently "npm install" command creats "/usr/etc" directory in
native sysroot. To prevent this $prefix's tail "/usr" is deleting
when creating "/etc" directory.

Testing Performed:
Locally built images does not contain /usr/etc directory.

:QA Notes:

:Issues Addressed:
[PLAT-14390] nodejs bypasses sstate and creates usr/etc directory in
native sysroot

Change-Id: I811f48f68b2990263c99a3e9dc51e714e5829b82
Reviewed-on: https://gpro.lgsvl.com/130995
Reviewed-by: Ievgen Sherstobitov <ievgen.sherstobitov@lge.com>
Tested-by: Ievgen Sherstobitov <ievgen.sherstobitov@lge.com>
Reviewed-by: Andrii Koval <andrii.koval@lge.com>
Reviewed-by: Tigran Avanesov <tigran.avanesov@lge.com>
---
 deps/npm/lib/config/core.js | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/deps/npm/lib/config/core.js b/deps/npm/lib/config/core.js
index b9851f9..6727009 100644
--- a/deps/npm/lib/config/core.js
+++ b/deps/npm/lib/config/core.js
@@ -147,6 +147,14 @@ function load_ (builtin, rc, cli, cb) {
     conf.once('load', afterUser)
   }
 
+  function cutTail(str, tail) {
+    var position = str.length-tail.length;
+    if (position>=0 && str.substring(position) === tail) {
+      return str.substring(0, position);
+    }
+    return str;
+  }
+
   function afterUser () {
     // globalconfig and globalignorefile defaults
     // need to respond to the 'prefix' setting up to this point.
@@ -154,7 +162,7 @@ function load_ (builtin, rc, cli, cb) {
     // return `~/local/etc/npmrc`
     // annoying humans and their expectations!
     if (conf.get('prefix')) {
-      var etc = path.resolve(conf.get('prefix'), 'etc')
+      var etc = path.resolve(cutTail(conf.get("prefix"), "/usr"), "etc")
       defaults.globalconfig = path.resolve(etc, 'npmrc')
       defaults.globalignorefile = path.resolve(etc, 'npmignore')
     }
